CONTRIBUTING.md

```diff
@@ -1,18 +1,139 @@
-# Contributing
+# Contributing to ICU4X Project
+
+`ICU4X` is an open source project and welcomes everyone to participate.
+
+In order to provide meaningful contributions, it is important to familiarize yourself with a set of documents which describe the [structure](./docs/charter.md) and [style](./docs/style-guide.md) used by the project.
+
+The list of [open issues](triaging.md) represent the current focus points of the project, and the [help wanted](https://github.com/unicode-org/icu4x/issues?q=is%3Aissue+label%3A%22help+wanted%22+no%3Aassignee+) lists the issues which we don't have resources to work on right now, but would consider accepting if someone volunteered to work on them.
+
+Issues are open to everyone to discuss and can be used to jump-start Pull Requests intended for the project.
+
+In most cases, the first step is to find or file a new issue related to your planned contribution, discuss it, and once you received a feedback indicating that the pull request would be welcomed you can start working on it.
+
+## Contributing a Pull Request
+
+The first step is to fork the repository to your namespace and create a branch off of the `master` branch to work with.
+
+That branch may end up containing one of more commits that are constituting the full scope of the pull request.
+
+### Single vs Multiple Commits per PR
```

Comment:
I've noticed in this PR that you're using a lot of force pushes. That makes it hard to tell what changed between reviews. I don't think that's a practice we should encourage.

This section is rather vague, but since it encourages a single commit per pull request, that implies that we endorse force pushes.

My preference would be to either remove this section entirely, or replace it with a discussion discouraging force pushes.

Reply:
Yeah, I've been experimenting with different models - in #142 I'm using additive commits, here I'm squashing.

The challenge here is that neither of those really work well for code reviews. Squashing means that you see one blob of changes that you have to comprehend and evaluate all at once.
Additive commits mean that you are either reviewing a "stream of thoughts" and 10 changes, or you look at them squashed anyway, so no benefit.

My aim with this paragraph is to encourage functional commits, where if the patch is small, you end up with a single commit per PR, and if the changes are large you end up with a patchset, in which every commit is functional and reviewable.

The latter model works really well at Mozilla, for example here you have a 13 parts patchset - https://bugzilla.mozilla.org/show_bug.cgi?id=1633935 or https://bugzilla.mozilla.org/show_bug.cgi?id=1623971 - that works well because it makes each patch easier to digest but also allows to distribute reviews among multiple people.

The issue is that maintaining a functional patchset in `git` is harder than in mercurial due to lack of `hg absorb`. That leads to people either adding additive changes or squashing.

The additive changes are a poor-mans-replacement for a patch history. In Phabricator, if you dig into any single commit out of the functional patchset, you can click on "History" and diff between revisions of that functional commit. Example: https://phabricator.services.mozilla.com/D77749?vs=on&id=288884#toc

Since git/github does not offer that, we're left with squashed single-commit PRs or additive commit PRs which only help to see the inter-diffs.

My goal here was to encourage functional patchsets, maybe via https://github.com/tummychow/git-absorb, but I can see how this is a pretty tall order to remodel around.

Maybe a solution would be to encourage functional or single commits as a state to aim for when requesting review (additive changes in the draft phase are useless), and then request additive commits in result of review rounds to allow reviewers to verify them (as some form of replacement for commit history)?

Reply:
Small PRs, or [multiple PRs chained together](https://github.com/ryanhiebert/probot-chain#note-you-probably-dont-need-this), can help with making them be more functional.

The single commit advice is fine I guess for the initial PR, e.g. after you turn it from Draft to Reviewing.  My main suggestion is that after a PR is open for review, it's much nicer if you can see diffs from commit to commit as the author is applying code review feedback.  So, basically what you said in the last paragraph in your reply.